receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,313 @@
+		StaticConfigs: []*targetgroup.Group{
+			{
+				Targets: []model.LabelSet{
+					{model.AddressLabel: model.LabelValue(fmt.Sprintf("localhost:%v", cfg.Port))},
+				},
+			},
+		},
+	}
+
+	return &prometheusreceiver.Config{
+		PrometheusConfig: &config.Config{
+			ScrapeConfigs: []*config.ScrapeConfig{scrapeConfig},
+		},
+	}
+}
+
+// getSubprocessConfig returns the subprocess config after the correct logic is made
+func getSubprocessConfig(cfg *Config, customName string) (*subprocessmanager.SubprocessConfig, error) {
+	if cfg.SubprocessConfig.Command == "" {
+		return nil, fmt.Errorf("no command to execute entered in config file for %v", cfg.Name())
+	}
+
+	subprocessConfig := &subprocessmanager.SubprocessConfig{}
+
+	subprocessConfig.Command = cfg.SubprocessConfig.Command
+	subprocessConfig.Env = cfg.SubprocessConfig.Env
+
+	return subprocessConfig, nil
+}
+
+// getCustomName will return the receiver's given custom name or try to generate one if none was given
+func getCustomName(cfg *Config) string {
+	// Try to get a custom name from the config (receivers should be named prometheus_exec/customName)
+	splitName := strings.SplitN(cfg.Name(), "/", 2)
+	if len(splitName) > 1 && splitName[1] != "" {
+		return splitName[1]
+	}
+	// fall back to the first part of the string, prometheus_exec, which should only happen if there is a single prometheus_exec receiver configured
+	return splitName[0]
+}
+
+// manageProcess will put the process in an infinite starting loop which goesd through the following steps
+// If the port is not defined by the user, one is generated and a new metrics receiver is built with the new port
+// All instances of {{port}} are replaced with the actual port (either defined by the user or generated)
+// Start the receiver if it is stopped (either first iteration or if it was previously shutdown)
+// We then start the subprocess, get its runtime and decide if the process is considered healthy or not -> the receiver is shutdown if it was deemed unhealthy
+// Finally, the wait time before the subprocess is restarted is computed and this goroutine sleeps for that amount of time, before restarting the loop from the start
+func (wrapper *prometheusReceiverWrapper) manageProcess() {
+	var (
+		elapsed       time.Duration
+		newPort       int
+		crashCount    int
+		err           error
+		subprocessErr error
+	)
+
+	for {
+
+		// Generate a port if none was specified and if process is unhealthy (Receiver has been stopped)
+		generatePort := wrapper.originalPort == 0 && elapsed <= healthyProcessTime
+		if generatePort {
+			newPort, err = wrapper.assignNewRandomPort(newPort)
+			if err != nil {
+				wrapper.logger.Info("assignNewRandomPort() error - killing this single process/receiver", zap.String("error", err.Error()))
+				return
+			}
+		}
+
+		// Replace the templating in the strings of the process data
+		wrapper.subprocessConfig = wrapper.fillPortPlaceholders(newPort)
```

Comment:
The logic here is a bit weird. It looks like this should be inside the `if` condition above since `newPort` will always be zero otherwise, but you actually check if the user specified a port inside `fillPortPlaceholders`. This is a little hard to follow. Can you just call `fillPortPlaceholders` with either `newPort` or `originalPort` here?

Reply:
I see what you mean! I changed it so that `newPort` is initialized to the same value as `originalPort` so the function is either called with the user defined value, or the randomly generated value if none was defined. Removed the logic within `fillPortPlaceholders`. thanks!